programs/stake/src/points.rs

```diff
@@ -115,7 +115,76 @@ pub(crate) fn calculate_stake_points_and_credits(
 ) -> CalculatedStakePoints {
     let credits_in_stake = stake.credits_observed;
     let credits_in_vote = new_vote_state.credits();
-    // if there is no newer credits since observed, return no point
+    // if there is no newer credits since observed, return zero points
+    if let Some(result) = zero_stake_points(
+        credits_in_vote,
+        credits_in_stake,
+        &inflation_point_calc_tracer,
+    ) {
+        return result;
+    }
+
+    let (points, new_credits_observed) = handle_epoch_credits(
+        new_vote_state.epoch_credits().iter().copied(),
+        stake,
+        stake_history,
+        new_rate_activation_epoch,
+        &inflation_point_calc_tracer,
+    );
+
+    CalculatedStakePoints {
+        points,
+        new_credits_observed,
+        force_credits_update_with_skipped_reward: false,
+    }
+}
+
+pub(crate) fn calculate_stake_points_and_credits_through_epoch(
+    max_epoch: Epoch,
+    stake: &Stake,
+    new_vote_state: &VoteState,
+    stake_history: &StakeHistory,
+    inflation_point_calc_tracer: Option<impl Fn(&InflationPointCalculationEvent)>,
+    new_rate_activation_epoch: Option<Epoch>,
+) -> CalculatedStakePoints {
+    let credits_in_stake = stake.credits_observed;
```

Comment:
Both this function and the previous function `fn calculate_stake_points_and_credits` have an implicit assumption that the passed `stake` hasn't had its credits updated yet because if it was updated already, we would return 0 below. This makes me wonder if this refactoring is a good idea at all. Because shouldn't we only be calling `calculate_stake_points_and_credits` with stake and vote state from the previous epoch? If we are calculating with state from a newer epoch than the epoch we are calculating points for, things could go pretty wrong. If we always use state from the previous epoch, the latest vote state epoch credits would be correct already and we wouldn't need any special handling for a prior epoch.

Reply:
> Both this function and the previous function fn calculate_stake_points_and_credits have an implicit assumption that the passed stake hasn't had its credits updated yet

I'm not totally sure what you mean by this. In fact, if the stake account *has* already had its credit updated (for example, when some partitions have already been distributed and then the partitions are recalculated with something like [my wip method](https://github.com/CriesofCarrots/solana/commit/6e98df733726ee4d6db487ed157ccb58d7a99c7d)), we want this method to return 0 so as not to double credit.

> Because shouldn't we only be calling calculate_stake_points_and_credits with stake and vote state from the previous epoch? If we are calculating with state from a newer epoch than the epoch we are calculating points for, things could go pretty wrong.

Definitely want to hear examples of how you envision this going wrong.
As I said, I think we want this function to respect the new stake-account state, if a reward has already been distributed. Meanwhile, vote accounts track credits by epoch, so the historical epoch data is still available even if newer epochs are in progress (up to the limit of the `epoch_credits` list, ofc).

I'm also not sure what alternative you are suggesting, really. Can you be more explicit?

Reply:
Main thing is that I didn't really understand how you're envisioning the partition recalculation to work yet. I was thinking we would do the exact same epoch boundary reward calculation whereas it seems you are working towards being able to basically recompute only the partitions that haven't been distributed yet?

One potential way this new calculation path could go wrong is that vote accounts could be withdrawn and deleted from the stakes cache after rewards calculation but before rewards distribution is finished. Of course vote accounts can only be closed if they haven't received credits for the prior epoch.. but there might be an edge case where if there was a small stake account which earned a small number of points from 2 epochs ago that previously got rounded to 0 and then in the current epoch the small number of points rounded to 1 lamport instead (due to slightly higher epoch point value), we would have a case where a stake account wouldn't receive anything on a restarted node because the vote account doesn't exist, but it would receive 1 lamport on a non-restarted node which correctly paid out a delayed small reward.

I just think that adding a new recalculation path is risky because vote state and stake state could have changed since the epoch boundary rewards calculation and future changes will have to keep this risk in mind in order to avoid introducing consensus bugs.